### ISSUE TITLE
Data type mismatches in model specification

### ISSUE THREAD

**Issue by ConnectedSystems** [open-AIMS/ADRIA.jl#603]
Model specification as generated by `ADRIA.model_spec(dom)` and loaded through `rs = load_results(path)` do not have the same data types for the same columns.

This leads to issues particularly for analyses which may expect Symbols vs Strings and vice versa (see for example: https://github.com/open-AIMS/ADRIA.jl/pull/601#discussion_r1403820435)

**Comment by ConnectedSystems** [comment]
Adding excerpts from the two model specifications.

Note the different data types

```julia
julia> ADRIA.model_spec(dom)[:, [:component, :fieldname, :val, :bounds, :default_bounds]]
253×5 DataFrame
 Row │ component           fieldname                          val         bounds                       default_bounds
     │ String              Symbol                             Real        Tuple…                       Tuple…
─────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
   1 │ EnvironmentalLayer  dhw_scenario                        1          (1.0, 51.0)                  (1.0, 51.0)
   2 │ EnvironmentalLayer  wave_scenario                       1          (1.0, 51.0)                  (1.0, 51.0)
   3 │ Intervention        guided                              0          (-1.0, 19.0)                 (-1.0, 19.0)

julia> rs.model_spec[:, [:component, :fieldname, :val, :bounds, :default_bounds]]
253×5 DataFrame
 Row │ component           fieldname                          val         bounds                             default_bounds
     │ String31            Symbol                             Float64     String                             String
─────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
   1 │ EnvironmentalLayer  dhw_scenario                        1.0        (1.0, 51.0)                        (1.0, 51.0)
   2 │ EnvironmentalLayer  wave_scenario                       1.0        (1.0, 51.0)                        (1.0, 51.0)
   3 │ Intervention        guided                              0.0        (-1.0, 19.0)                       (-1.0, 19.0)
```

**Comment by ConnectedSystems** [comment]
> what do you think about having one row with types for each col inside model_spec.csv ? Or do you have any other ideas?

Yes, that's one approach - to list the types in a comment directly above the column headers and make that part of the specification.

Alternatively, switch to Zarr format instead of CSV and put the metadata of column types somewhere in Zarr.

Or specify it in the ResultSet metadata (inside the datapackage.json file?)...

Personally I like the Zarr-based solution...

**Comment by Zapiano** [comment]
I also like the Zarr-based solution. I don't know what are the complications (can you elaborate?) but one problem I see is that is it harder to implement than just adding a commented row to the csv.

**Comment by ConnectedSystems** [comment]
> harder to implement than just adding a commented row to the csv

Yes easier, but doing that would open a whole lot of security implications. Because we'd be evaluating string as code to obtain the equivalent data types.

Actually, this is true for all the ideas above (but somewhat less so for Zarr).

Maybe I'm overthinking it. Let's sleep on it.

**Comment by Zapiano** [comment]
> > harder to implement than just adding a commented row to the csv
> 
> Yes easier, but doing that would open a whole lot of security implications. Because we'd be evaluating string as code to obtain the equivalent data types.
> 
> Actually, this is true for all the ideas above (but somewhat less so for Zarr).
> 
> Maybe I'm overthinking it. Let's sleep on it.

I was thinking: is it possible to just pass `dom` or, at least, a vector of types, as an argument to `load_results`? This function is being called in a lot of places so I'm not completely sure, but I see that two main places are inside [`combine_results`](https://github.com/open-AIMS/ADRIA.jl/blob/0667a6b6235bc2eeb7e7cc2d85bd96296c1b2b9e/src/io/ResultSet.jl#L267) and inside [`run_scenarios`](https://github.com/open-AIMS/ADRIA.jl/blob/0667a6b6235bc2eeb7e7cc2d85bd96296c1b2b9e/src/scenario.jl#L201). I didn't find any call to `combine_results` inside ADRIA, how/where/when is this being used? But about `run_scenarios`, we have the domain there, we could pass the types as arguments to `load_results`.

**Comment by ConnectedSystems** [comment]
> I was thinking: is it possible to just pass dom or, at least, a vector of types, as an argument to load_results? 

Do you remember what the motivation is behind keeping the Domain and ResultSets separate?

**Comment by Zapiano** [comment]
> > I was thinking: is it possible to just pass dom or, at least, a vector of types, as an argument to load_results?
> 
> Do you remember what the motivation is behind keeping the Domain and ResultSet separate?

Yes, you're right, sorry. Stupid idea.

**Comment by Zapiano** [comment]
> Maybe, if we're willing to drop support for any prior versions, the simplest solution is to take the expected types from within ADRIA itself.

I think now is a good time to implement changes that are important like this but have the cost of dropping support to prior versions. Soon we will release 1.0.0 (right?) anyway. 

That said, what do you mean by "take the types from within ADRIA"? What I like about having the types saved somewhere within result set is that it allows us support other models/domains in the future. So far I'm leaning towards the Zarr-based solution. 

> > I didn't find any call to combine_results inside ADRIA, how/where/when is this being used?
> 
> This was the first thing you worked on. I said not to delete the function because we still want to be able to merge ResultSets. Later we want to be able to add to an existing ResultSet and other things too.

Ok, I remember that.

**Comment by ConnectedSystems** [comment]
> That said, what do you mean by "take the types from within ADRIA"?

The model specification is defined by the composition of the various components (Intervention, CriteriaWeight, EnvironmentalLayer, Coral). 

I think only EnvironmentalLayer requires Domain metadata to be created (e.g., the number of available trajectories), but have to check this.

Given we only want the data types, I guess it should be possible to create a model spec with just the default values and infer the types from that.

The Zarr-based approach has the same issue where we'd have to infer the correct types from a metadata string - running strings as code should be avoided. But again, maybe I'm overthinking it.

**Comment by ConnectedSystems** [comment]
Another option might be to store a `serialized` copy of the DataFrame. 

This might be the simplest approach but makes it susceptible to:

1. Changes to Julia
2. Changes to the DataFrames package

If we're dropping backward and forward compatibility (i.e., users are expected to run exactly the same versions for everything) then it might be acceptable.
